File: src/events.ts
export type Evidence = "provider_reported" | "runtime_reported" | "tokenizer_counted" | "heuristic_estimate" | "reconstructed_baseline" | "controlled_experiment";
export type ExecutionLocation = "local" | "remote" | "unknown";
export type TransformScope = "tool_output" | "request" | "attempt" | "task" | "experiment";
export type UsageCategory = "primary" | "auxiliary" | "restoration";

export interface EventBase {
  schemaVersion: 1;
  eventId: string;
  occurredAt: string;
  sourceId: string;
  sourceEventId: string;
  projectId: string;
  taskId: string | null;
  sessionId: string;
  requestId: string | null;
  attemptId: string | null;
  clientId: string;
  adapterVersion: string;
  modelId: string | null;
  providerId: string | null;
  executionLocation: ExecutionLocation;
  evidence: Evidence;
}
export type EconomyEvent = EventBase & (
  | { eventType: "transform"; payload: { scope: TransformScope; inputTokens: number | null; outputTokens: number | null; tokenizer: string | null; transformId: string; parentTransformId: string | null; inputArtifactId: string | null; outputArtifactId: string | null; strategy: string; sent: boolean } }
  | { eventType: "usage"; payload: { inputTokens: number | null; outputTokens: number | null; cacheReadTokens: number | null; cacheWriteTokens: number | null; complete: boolean; category: UsageCategory; callId: string | null; costUsd: number | null; observationScope?: 'call' | 'session_delta'; rawUsage?: Record<string, unknown>; normalizerVersion?: string; costProvenance?: { priceSource: string; priceDate: string; priceModelId: string } | null } }
  | { eventType: "decision"; payload: { decision: string; reason: string | null; strategy: string | null } }
  | { eventType: "outcome"; payload: { outcome: string; reason: string | null } }
  | { eventType: "reconciliation"; payload: { targetEventId: string; effectiveInputTokens: number | null; effectiveOutputTokens: number | null; effectiveCacheReadTokens: number | null; effectiveCacheWriteTokens: number | null; effectiveCostUsd: number | null; effectiveCostProvenance?: { priceSource: string; priceDate: string; priceModelId: string } | null; effectiveComplete?: boolean; reason: string } }
);

const nonempty = (v: unknown): v is string => typeof v === "string" && v.trim().length > 0;
const nullableString = (v: unknown): v is string | null => v === null || nonempty(v);
const count = (v: unknown): v is number | null => v === null || (typeof v === "number" && Number.isSafeInteger(v) && v >= 0);
const money = (v: unknown): v is number | null => v === null || (typeof v === "number" && Number.isFinite(v) && v >= 0);
const validCostProvenance = (v: unknown): boolean => record(v) && nonempty(v.priceSource) && /^\d{4}-\d{2}-\d{2}$/.test(String(v.priceDate)) && nonempty(v.priceModelId);
const record = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);
const utc = (v: unknown): v is string => typeof v === "string" && /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/.test(v) && !Number.isNaN(Date.parse(v)) && new Date(v).toISOString() === v;
export function normalizeUtc(value: string): string { const date = new Date(value); if (!Number.isFinite(date.getTime())) throw new Error("Invalid UTC date"); return date.toISOString(); }

export function validateEvent(input: unknown): EconomyEvent {
  if (!record(input)) throw new Error("Event must be an object");
  if (Buffer.byteLength(JSON.stringify(input)) > 65536) throw new Error('Event exceeds 64 KiB limit');
  const e = input;
  if (e.schemaVersion !== 1 || !nonempty(e.eventId) || !utc(e.occurredAt) || !nonempty(e.sourceId) || !nonempty(e.sourceEventId) || !nonempty(e.projectId)) throw new Error("Invalid event envelope");
  for (const key of ["sessionId", "clientId", "adapterVersion"]) if (!nonempty(e[key])) throw new Error(`Invalid ${key}`);
  for (const key of ["taskId", "requestId", "attemptId", "modelId", "providerId"]) if (!nullableString(e[key])) throw new Error(`Invalid ${key}`);
  if (!["local", "remote", "unknown"].includes(String(e.executionLocation))) throw new Error("Invalid execution location");
  if (!["provider_reported", "runtime_reported", "tokenizer_counted", "heuristic_estimate", "reconstructed_baseline", "controlled_experiment"].includes(String(e.evidence))) throw new Error("Invalid evidence");
  if (!record(e.payload)) throw new Error("Invalid payload");
  const p = e.payload;
  switch (e.eventType) {
    case "transform":
      if (!["tool_output", "request", "attempt", "task", "experiment"].includes(String(p.scope)) || !count(p.inputTokens) || !count(p.outputTokens) || !nullableString(p.tokenizer) || !nonempty(p.transformId) || !nullableString(p.parentTransformId) || !nullableString(p.inputArtifactId) || !nullableString(p.outputArtifactId) || !nonempty(p.strategy) || typeof p.sent !== "boolean") throw new Error("Invalid transform payload");
      break;
    case "usage":
      if (![p.inputTokens, p.outputTokens, p.cacheReadTokens, p.cacheWriteTokens].every(count) || typeof p.complete !== "boolean" || !["primary", "auxiliary", "restoration"].includes(String(p.category)) || !nullableString(p.callId) || !money(p.costUsd) || (p.costUsd !== null && (!nonempty(e.modelId) || !validCostProvenance(p.costProvenance)))) throw new Error("Invalid usage payload");
      if (p.observationScope !== undefined && p.observationScope !== 'call' && p.observationScope !== 'session_delta') throw new Error('Invalid observation scope');
      if (p.rawUsage !== undefined && !record(p.rawUsage)) throw new Error('Invalid original usage counters');
      if (p.normalizerVersion !== undefined && !nonempty(p.normalizerVersion)) throw new Error('Invalid normalizer version');
      if (p.inputTokens !== null && p.cacheReadTokens !== null && Number(p.cacheReadTokens) > Number(p.inputTokens)) throw new Error('Cache reads exceed normalized input');
      break;
    case "decision":
      if (!nonempty(p.decision) || !nullableString(p.reason) || !nullableString(p.strategy)) throw new Error("Invalid decision payload");
      break;
    case "outcome":
      if (!nonempty(p.outcome) || !nullableString(p.reason)) throw new Error("Invalid outcome payload");
      break;
    case "reconciliation":
      if (!nonempty(p.targetEventId) || ![p.effectiveInputTokens, p.effectiveOutputTokens, p.effectiveCacheReadTokens, p.effectiveCacheWriteTokens].every(count) || !money(p.effectiveCostUsd) || (p.effectiveCostUsd !== null && (!nonempty(e.modelId) || !validCostProvenance(p.effectiveCostProvenance))) || (p.effectiveComplete !== undefined && typeof p.effectiveComplete !== "boolean") || !nonempty(p.reason)) throw new Error("Invalid reconciliation payload");
      break;
    default: throw new Error("Unknown event type");
  }
  return input as unknown as EconomyEvent;
}

File: python/julius/events.py
"""Versioned, strict event schema for the append-only economy ledger."""

from __future__ import annotations

import json
import re
from datetime import datetime, timezone
from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, field_validator, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")


def _nonempty(value: str) -> str:
    if not value.strip():
        raise ValueError("Identifier must be nonempty")
    return value


def normalize_utc(value: str) -> str:
    try:
        instant = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as exc:
        raise ValueError("Invalid UTC date") from exc
    if instant.tzinfo is None:
        raise ValueError("Timestamp must include timezone")
    return (
        instant.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
    )


class PriceProvenance(StrictModel):
    priceSource: str
    priceDate: str
    priceModelId: str

    @field_validator("priceSource", "priceModelId")
    @classmethod
    def names(cls, value: str) -> str:
        return _nonempty(value)

    @field_validator("priceDate")
    @classmethod
    def date(cls, value: str) -> str:
        if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", value):
            raise ValueError("Invalid price date")
        datetime.fromisoformat(value)
        return value


TokenCount = Annotated[int, Field(ge=0, le=9007199254740991)]
Money = Annotated[float, Field(ge=0, allow_inf_nan=False)]


class TransformPayload(StrictModel):
    scope: Literal["tool_output", "request", "attempt", "task", "experiment"]
    inputTokens: TokenCount | None
    outputTokens: TokenCount | None
    tokenizer: str | None
    transformId: str
    parentTransformId: str | None
    inputArtifactId: str | None
    outputArtifactId: str | None
    strategy: str
    sent: bool

    @field_validator("transformId", "strategy")
    @classmethod
    def names(cls, value: str) -> str:
        return _nonempty(value)


class UsagePayload(StrictModel):
    inputTokens: TokenCount | None
    outputTokens: TokenCount | None
    cacheReadTokens: TokenCount | None
    cacheWriteTokens: TokenCount | None
    complete: bool
    category: Literal["primary", "auxiliary", "restoration"]
    callId: str | None
    costUsd: Money | None
    observationScope: Literal["call", "session_delta"] | None = None
    rawUsage: dict[str, Any] | None = None
    normalizerVersion: str | None = None
    costProvenance: PriceProvenance | None = None

    @model_validator(mode="after")
    def valid_cost(self) -> UsagePayload:
        if self.costUsd is not None and self.costProvenance is None:
            raise ValueError("Known cost requires provenance")
        if (
            self.inputTokens is not None
            and self.cacheReadTokens is not None
            and self.cacheReadTokens > self.inputTokens
        ):
            raise ValueError("Cache reads exceed input")
        return self


class DecisionPayload(StrictModel):
    decision: str
    reason: str | None
    strategy: str | None


class OutcomePayload(StrictModel):
    outcome: str
    reason: str | None


class ReconciliationPayload(StrictModel):
    targetEventId: str
    effectiveInputTokens: TokenCount | None
    effectiveOutputTokens: TokenCount | None
    effectiveCacheReadTokens: TokenCount | None
    effectiveCacheWriteTokens: TokenCount | None
    effectiveCostUsd: Money | None
    effectiveCostProvenance: PriceProvenance | None = None
    effectiveComplete: bool | None = None
    reason: str

    @model_validator(mode="after")
    def valid_cost(self) -> ReconciliationPayload:
        if self.effectiveCostUsd is not None and self.effectiveCostProvenance is None:
            raise ValueError("Known corrected cost requires provenance")
        return self


class EventBase(StrictModel):
    schemaVersion: Literal[1]
    eventId: str
    occurredAt: str
    sourceId: str
    sourceEventId: str
    projectId: str
    taskId: str | None
    sessionId: str
    requestId: str | None
    attemptId: str | None
    clientId: str
    adapterVersion: str
    modelId: str | None
    providerId: str | None
    executionLocation: Literal["local", "remote", "unknown"]
    evidence: Literal[
        "provider_reported",
        "runtime_reported",
        "tokenizer_counted",
        "heuristic_estimate",
        "reconstructed_baseline",
        "controlled_experiment",
    ]

    @field_validator(
        "eventId",
        "sourceId",
        "sourceEventId",
        "projectId",
        "sessionId",
        "clientId",
        "adapterVersion",
    )
    @classmethod
    def identifiers(cls, value: str) -> str:
        return _nonempty(value)

    @field_validator("occurredAt")
    @classmethod
    def timestamp(cls, value: str) -> str:
        if (
            not re.fullmatch(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z", value)
            or normalize_utc(value) != value
        ):
            raise ValueError("Canonical UTC milliseconds required")
        return value

    @model_validator(mode="after")
    def model_cost(self) -> EventBase:
        payload = getattr(self, "payload", None)
        if payload is not None:
            known_cost = (
                getattr(payload, "costUsd", None) is not None
                or getattr(payload, "effectiveCostUsd", None) is not None
            )
            if known_cost and self.modelId is None:
                raise ValueError("Known cost requires known model")
            provenance = getattr(payload, "costProvenance", None) or getattr(
                payload, "effectiveCostProvenance", None
            )
            if provenance is not None and provenance.priceModelId != self.modelId:
                raise ValueError("Price model does not match event model")
        return self


class TransformEvent(EventBase):
    eventType: Literal["transform"]
    payload: TransformPayload


class UsageEvent(EventBase):
    eventType: Literal["usage"]
    payload: UsagePayload


class DecisionEvent(EventBase):
    eventType: Literal["decision"]
    payload: DecisionPayload


class OutcomeEvent(EventBase):
    eventType: Literal["outcome"]
    payload: OutcomePayload


class ReconciliationEvent(EventBase):
    eventType: Literal["reconciliation"]
    payload: ReconciliationPayload


EconomyEvent = Annotated[
    Union[TransformEvent, UsageEvent, DecisionEvent, OutcomeEvent, ReconciliationEvent],
    Field(discriminator="eventType"),
]
_event_adapter: TypeAdapter[EconomyEvent] = TypeAdapter(EconomyEvent)


def validate_event(value: Any) -> dict[str, Any]:
    if len(json.dumps(value, ensure_ascii=False).encode("utf-8")) > 65536:
        raise ValueError("Event exceeds 64 KiB limit")
    event = _event_adapter.validate_python(value, strict=True)
    return event.model_dump(exclude_none=False)
